### PR TITLE
Studio-NDTiffAdapter: Simplifies function getImagesMatching()

### LINKED ISSUE #1839: java.lang.OutOfMemoryError: Java heap space, when displaying (open virtual) large (>20gb) ndtiff stores (recordings)
Hi!
When opening large ndtiff recordings (1000 frames 5000x5000px) or creating a display (createDisplay(ndtiffstore)) for the ndtiffstore a lot of memory is needed, or the system runs out of memory.

Example:
Record 1000 frames with Camera-OnCameraCCDXSize:5120, Camera-OnCameraCCDYSize:5120 (MMConfig_Demo.cfg). When trying to open with open(virtual) the system runs out of memory (>20Gb memory usage).

```text
[UI]  "Open" clicked in JDialog.
2023-11-16T14:09:25.184986 tid6000 [IFO,App] Class: class org.micromanager.display.internal.displaywindow.DisplayController
2023-11-16T14:09:25.185022 tid6000 [IFO,App] Classloader: ij.io.PluginClassLoader@23594b45
2023-11-16T14:09:29.855803 tid8216 [dbg,App] [UI] JMenu "File" clicked in MainFrame.
2023-11-16T14:09:30.223334 tid6000 [IFO,App] Thread 150 (Thread-8) terminated with uncaught exception
2023-11-16T14:09:30.223367 tid6000 [IFO,App] java.lang.OutOfMemoryError: Java heap space
2023-11-16T14:09:30.223511 tid6000 [IFO,App]   at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:57)
2023-11-16T14:09:30.223514 tid6000 [IFO,App]   at java.nio.ByteBuffer.allocate(ByteBuffer.java:335)
2023-11-16T14:09:30.223517 tid6000 [IFO,App]   at org.micromanager.ndtiffstorage.NDTiffReader.readTaggedImage(NDTiffReader.java:299)
2023-11-16T14:09:30.223521 tid6000 [IFO,App]   at org.micromanager.ndtiffstorage.NDTiffReader.readImage(NDTiffReader.java:173)
2023-11-16T14:09:30.223522 tid6000 [IFO,App]   at org.micromanager.ndtiffstorage.ResolutionLevel.getImage(ResolutionLevel.java:178)
2023-11-16T14:09:30.223524 tid6000 [IFO,App]   at org.micromanager.ndtiffstorage.NDTiffStorage.getImage(NDTiffStorage.java:996)
2023-11-16T14:09:30.223525 tid6000 [IFO,App]   at org.micromanager.ndtiffstorage.NDTiffStorage.getImage(NDTiffStorage.java:943)
2023-11-16T14:09:30.223529 tid6000 [IFO,App]   at org.micromanager.data.internal.ndtiff.NDTiffAdapter$1.apply(NDTiffAdapter.java:244)
2023-11-16T14:09:30.223530 tid6000 [IFO,App]   at org.micromanager.data.internal.ndtiff.NDTiffAdapter$1.apply(NDTiffAdapter.java:241)
2023-11-16T14:09:30.223531 tid6000 [IFO,App]   at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
2023-11-16T14:09:30.223534 tid6000 [IFO,App]   at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
2023-11-16T14:09:30.223536 tid6000 [IFO,App]   at java.util.concurrent.ConcurrentHashMap$KeySpliterator.forEachRemaining(ConcurrentHashMap.java:3527)
2023-11-16T14:09:30.223537 tid6000 [IFO,App]   at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
2023-11-16T14:09:30.223540 tid6000 [IFO,App]   at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
2023-11-16T14:09:30.223542 tid6000 [IFO,App]   at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
2023-11-16T14:09:30.223544 tid6000 [IFO,App]   at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2023-11-16T14:09:30.223547 tid6000 [IFO,App]   at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
2023-11-16T14:09:30.223549 tid6000 [IFO,App]   at org.micromanager.data.internal.ndtiff.NDTiffAdapter.getImagesMatching(NDTiffAdapter.java:248)
2023-11-16T14:09:30.223550 tid6000 [IFO,App]   at org.micromanager.data.internal.ndtiff.NDTiffAdapter.getImagesIgnoringAxes(NDTiffAdapter.java:256)
2023-11-16T14:09:30.223553 tid6000 [IFO,App]   at org.micromanager.data.internal.DefaultDatastore.getImagesIgnoringAxes(DefaultDatastore.java:204)
2023-11-16T14:09:30.223555 tid6000 [IFO,App]   at org.micromanager.display.internal.displaywindow.DisplayController.handleDisplayPosition(DisplayController.java:561)
2023-11-16T14:09:30.223558 tid6000 [IFO,App]   at org.micromanager.display.AbstractDataViewer.setDisplayPosition(AbstractDataViewer.java:229)
2023-11-16T14:09:30.223560 tid6000 [IFO,App]   at org.micromanager.display.AbstractDataViewer.setDisplayPosition(AbstractDataViewer.java:242)
2023-11-16T14:09:30.223561 tid6000 [IFO,App]   at org.micromanager.display.internal.displaywindow.DisplayController.create(DisplayController.java:241)
2023-11-16T14:09:30.223565 tid6000 [IFO,App]   at org.micromanager.display.internal.displaywindow.DisplayController.access$000(DisplayController.java:86)
2023-11-16T14:09:30.223566 tid6000 [IFO,App]   at org.micromanager.display.internal.displaywindow.DisplayController$Builder.build(DisplayController.java:211)
2023-11-16T14:09:30.223569 tid6000 [IFO,App]   at org.micromanager.display.internal.DefaultDisplayManager.loadDisplays(DefaultDisplayManager.java:364)
2023-11-16T14:09:30.223571 tid6000 [IFO,App]   at org.micromanager.internal.menus.FileMenu.lambda$promptToOpenFile$4(FileMenu.java:117)
2023-11-16T14:09:30.223573 tid6000 [IFO,App]   at org.micromanager.internal.menus.FileMenu$$Lambda$233/1515206427.run(Unknown Source)
2023-11-16T14:09:30.223576 tid6000 [IFO,App]   at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by nicost**
so that it does no longer go through the complete dataset.  This may lead to problems in certain cases, but at least makes virtual datasets readable again.
Closes #1839

### CODE REVIEW

**Review by marktsuchida — COMMENTED** [review]
Just to check, is the intent not to change what `getImagesMatching()` returns, but to change how it computes the matching set?

> This may lead to problems in certain cases

Which cases?

**Comment by nicost** [comment]
I believe it now only returns an exact match, so no longer the super set (i.e., if one only asks for channel == 1, it should return all images with channel == 1).  I do not know how to implement that behavior with the NDTiffStorage back-end.

**Comment by henrypinkard** [comment]
So the problem was that it was too slow? 

> I believe it now only returns an exact match, so no longer the super set (i.e., if one only asks for channel == 1, it should return all images with channel == 1). I do not know how to implement that behavior with the NDTiffStorage back-end.

Its not clear to me that this would cause problems (depends on studio functionality i think).

If you wanted to keep the same logic but improve speed I believe there's a parallel version of stream.map

But either way is fine by me

**Review by henrypinkard — APPROVED** [review]

File: mmstudio/src/main/java/org/micromanager/data/internal/ndtiff/NDTiffAdapter.java [hunk 1]
```diff
@@ -222,30 +222,28 @@ public Iterable<Coords> getUnorderedImageCoords() {
...
-            } else if ((Integer) stringIntegerHashMap.get(axis) != coords.getIndex(axis)) {
-               return false;
-            }
-         }
-         return true;
-      });
-      return axesStream.map(new Function<HashMap<String, Object>, Image>() {
-         @Override
-         public Image apply(HashMap<String, Object> axes) {
-            TaggedImage ti = addEssentialImageMetadata(storage_.getImage(axes), axes);
```
Comment (henrypinkard) [inline mmstudio/src/main/java/org/micromanager/data/internal/ndtiff/NDTiffAdapter.java:244]: This is the line that I think would need to change

**Comment by nicost** [comment]
> This is the line that I think would need to change

To what? I do not really understand this code.

**Comment by henrypinkard** [comment]
from `storage_.getImage` to `storage_.getImageMetadata` (so that it doesnt read pixels when it doesnt need them)

and then change the `addEssentialImageMetadata` to do whatever it is the studio needs

**Comment by nicost** [comment]
What I would need from the storage class is a method that tells which axes are available and which images (with axes coordinates) are in the dataset.  It it unclear from the NDTiff Storage API is such a method exits, and documentation such as:

 ```
  /**
    * Get the metadata for a single image
    *
    * @param axes
    * @return
    */
   public JSONObject getImageMetadata(HashMap<String, Object> axes);
```

is not very helpful.  I will merge this PR since it at least improves the situation, but welcome work on making this actually work right.

**Pull request merged**